### ISSUE TITLE
reportOptionalMemberAccess: On which side of `is` operator a `None` literal appears affects narrowing of `Optional` type.

### ISSUE THREAD

**Issue by emcd** [microsoft/pyright#10145]
**Describe the bug**

Although uncommon in Python, `if None is <variable>: return` is semantically valid and should be equivalent to `if <variable> is None: return`. However, Pyright does not properly narrow the type of `<variable>` in the uncommon case, which leads to false positive `reportOptionalMemberAccess` diagnostics.

**Code or Screenshots**

Minimal Reproducer:

```python
def process_text_good(value: str | None) -> str:
    # No bug in this function
    if value is None:
        return "default"
    # Pyright recognizes `value` can only be a string at this point
    return value.upper()


def process_text_bad(value: str | None) -> str:
    # Using `None is value` triggers the bug
    if None is value:
        return "default"
    # Pyright incorrectly thinks `value` could still be None here
    return value.upper()  # Should be recognized as str, but Pyright flags this
```

Error:
```text
/home/me/src/python-mimeogram/bugs/pyright-lvalue.py
  /home/me/src/python-mimeogram/bugs/pyright-lvalue.py:16:18 - error: "upper" is not a known attribute of "None" (reportOptionalMemberAccess)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
```text
$ hatch --env develop run pyright --version
pyright 1.1.397
```

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug. The `x is None` expression form is a supported type guard pattern, whereas `None is x` is not. See [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) for a full list of supported type guard patterns.

We do occasionally add support for new patterns, but each one requires custom logic and can negatively affect overall type analysis performance, so the bar is high. I would need to see good evidence (e.g. in the form of upvotes from other pyright users) that this is a common pattern. I personally haven't seen `None is x` used in real-world code. I also find this more difficult to read, so it's not surprising for `x is None` to be preferred.

**Comment by emcd** [comment]
I doubt that it is common, but it does sometimes fall out of refactors when code is shuffled around on lines. Also, some of us have been around long enough to have been scarred by languages where using constant l-values is a defensive programming practice and have carried those habits to languages which do not need them. (I'm one of these people and almost certainly a minority.)

Thanks for the link.
